src/loader/BlendLoader.py

```diff
@@ -42,59 +75,164 @@ class BlendLoader(LoaderInterface):
+                            the .blend file version supports {} ID. Current \
+                            Blender API Version {}".format(
+                    data_block_name, data_block_name, bpy.app.version_string))
+
+        return blend_file_datablock_names[index]
+
+    def _set_datablock_properties(self, resource: bpy.types.ID):
+        """
+        Sets the custom properties of **non object** resources like materials,
+        textures, images, etc.
+
+        Note: Some datablock types like bpy.types.Light, bpy.types.Mesh, bpy.types.Camera etc
+        are wrapped in bby.types.Object which act as a container of these object. In that case
+        setting the properties of the container object does not set the properties of underlying datablock like
+        camera and vice versa. Setting the bpy.data.objects["Light"] and bpy.data.lights["light"] can
+        have different properties. This function sets properties of all types materials, lights,
+        cameras even if they are loaded as an object. For Objects that wraps Meshes use  self._set_properties
+        instead
+
+        :param resource: The datablock for which the properties are set
+        """
+        properties = self.config.get_raw_dict("add_properties", {})
+        for key, value in properties.items():
+            if key.startswith("cp_"):
+                key = key[3:]
+                resource[key] = value
             else:
-                raise Exception("Unsupported datablock/folder name: " + load_from +
-                                "\nSupported names: " + str(self.known_datablock_names.keys()) +
-                                "\nIf your ID exists, but not supported, please append a new pair of "
-                                "{type ID(folder name): parameter name} to the 'known_datablock_names' dict. Use this "
-                                "for finding your parameter name: " + str(dir(data_from)))
-        newly_loaded_objects = list(set(bpy.context.scene.objects) - previously_loaded_objects)
-        self._set_properties(newly_loaded_objects)
+                raise RuntimeError(
+                    "Loader modules support setting only custom properties. Use 'cp_' prefix for keys. "
+                    "Use manipulators.Entity for setting object's attribute values.")
+            resource[key] = value
+
+    def _get_camera_keyframes(self, camera):
+        """
+        Get Keyframes from animation data of a Camera Object.
+
+        :param camera: The camera for which the keyframes are extracted
+        :return: [] kerframes
+        """
+        keyframes = []
+        anim = camera.animation_data
+        if anim is not None and anim.action is not None:
+            for fcu in anim.action.fcurves:
+                for keyframe in fcu.keyframe_points:
+                    x, y = keyframe.co
+                    if x not in keyframes:
+                        keyframes.append((math.ceil(x)))
```

Comment:
Why is this `math.ceil` necessary?

Reply:
Frame numbers are in range (float [-inf, inf]) so a variable can have potentially infinite values between between two frames. We need unique frames so round off the frames.

Reply:
details are documented inline.